### PR TITLE
#478 Corrected trace bug for Windows because of singleton non-unique.…

### PR DESCRIPTION
**Pull request by gautierbureau**
… CMake corrections of dependencies problems.

### CODE REVIEW

File: dynawo/3rdParty/CMakeLists.txt [hunk 1]
```diff
@@ -538,6 +538,45 @@ else()
   set(SUITESPARSE_HOME ${install_dir})
 endif()
 
+if(UNIX)
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/3rdParty/CMakeLists.txt]: agree they are different topic but this part is actually the only new piece of code regarding nicslu and all cmake are working so separating them would not be very productive

File: dynawo/3rdParty/sundials/patch/common/sundials.patch [hunk 2]
```diff
@@ -1,28 +1,26 @@
...
-+++ b/include/ida/ida.h
-@@ -134,6 +134,7 @@ SUNDIALS_EXPORT int IDASetUserData(void *ida_mem, void *user_data);
+diff '--exclude=.git' -urpN sundials-4.1.0/include/ida/ida.h sundials-4.1.0.patch/include/ida/ida.h
+--- sundials-4.1.0/include/ida/ida.h	2019-02-12 19:51:21.000000000 +0100
++++ sundials-4.1.0.patch/include/ida/ida.h	2020-01-30 08:53:26.672843634 +0100
+@@ -134,6 +134,7 @@ SUNDIALS_EXPORT int IDASetUserData(void
  SUNDIALS_EXPORT int IDASetMaxOrd(void *ida_mem, int maxord);
  SUNDIALS_EXPORT int IDASetMaxNumSteps(void *ida_mem, long int mxsteps);
  SUNDIALS_EXPORT int IDASetInitStep(void *ida_mem, realtype hin);
 +SUNDIALS_EXPORT int IDASetMinStep(void *ida_mem, realtype hmin);
```
[the remark this answers is not in the document]
Comment (gautierbureau) [reply dynawo/3rdParty/sundials/patch/common/sundials.patch]: nothing, I just added a trick in the comments to avoid the patch to be appliable with two consecutives run

File: dynawo/3rdParty/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/3rdParty/CMakeLists.txt]: I disagree with this as:
a. You will need to test compilation and trace generation again anyway due to the addition of a local copy of IODicos and Trace getter handles;
b. This commit is quite big as it is a melting pot of many little fixes, fix of coding style and two bigger modifications (fix of the bug of the trace and nicslu restoration). There is no ticket to trace most of those modifications, which is not a big deal when it is not affecting the functionalities, but it is not the case of nicslu fix. In addition, nicslu fix is not clearly mentioned in the commit message, meaning that there is no trace of this correction neither in the ticket system nor in the commit history. This is not good for maintenance and transparency;
c. I just double-checked and the multiple new pieces of code related to nicslu are easy to identify (always surrounded by a test on NICSLU_FOUND) and independant from the rest. I am not asking to filter the others modifications in the cmakes (in particular the sundials ones), I just would like to have an independant commit with only the nicslu related stuff so that it is easy to identify and visible. 
i hope this convinces you, otherwise I am available to discuss this anytime.

File: dynawo/3rdParty/sundials/patch/common/sundials.patch (hunk 2 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/3rdParty/sundials/patch/common/sundials.patch]: OK

File: dynawo/sources/Common/test/TestIoDico.cpp [hunk 3]
```diff
@@ -32,14 +33,14 @@ TEST(CommonIoDicoTest, testCommonIoDicosTest) {
   ASSERT_EQ(IoDicos::hasIoDico("MyIoDico"), false);
   ASSERT_THROW(IoDicos::getIoDico("MyIoDico"), MessageError);
 
-  boost::shared_ptr<IoDicos> dicos = IoDicos::getInstance();
-  dicos->addPath("res");
-  ASSERT_NO_THROW(dicos->addDico("MyIoDico", "dico", ""));
-  ASSERT_THROW(dicos->addDico("MyIoDico", "", ""), MessageError);
-  ASSERT_THROW(dicos->addDico("MyIoDico", "MyDummyDico", ""), MessageError);
+  IoDicos& dicos = IoDicos::getInstance();
```
Comment (rosiereflo) [inline dynawo/sources/Common/test/TestIoDico.cpp]: Is there a way to unit test that the Trace and Dicos are always unique? Maybe by compiling and loading a dummy library that writes something in the log and make sure it ends up in the proper file? Might be the subject of another commit though.
